Sources/Testing/Traits/Trait.swift

```diff
@@ -41,6 +41,125 @@ public protocol Trait: Sendable {
+  /// at runtime, even if the test has the trait applied. This can improve
+  /// performance and make diagnostics clearer by avoiding an unnecessary call
+  /// to ``TestExecuting/execute(_:for:testCase:)``.
+  ///
+  /// If this trait's type does not conform to ``TestExecuting`` and its
+  /// associated ``Trait/TestExecutor`` type is the default `Never`, then this
+  /// method returns `nil` by default. This means that instances of this type
+  /// will not perform any custom test execution for tests they are applied to.
+  func executor(for test: Test, testCase: Test.Case?) -> TestExecutor?
+}
+
+/// A protocol that allows customizing the execution of a test function (and
+/// each of its cases) or a test suite by performing custom code before or after
+/// it runs.
+///
+/// Types conforming to this protocol may be used in conjunction with a
+/// ``Trait``-conforming type by implementing the
+/// ``Trait/executor(for:testCase:)-26qgm`` method, allowing custom traits to
+/// customize the execution of tests. Consolidating common set-up and tear-down
+/// logic for tests which have similar needs allows each test function to be
+/// more succinct with less repetitive boilerplate so it can focus on what makes
+/// it unique.
+public protocol TestExecuting: Sendable {
+  /// Execute a function for the specified test and/or test case.
+  ///
+  /// - Parameters:
+  ///   - function: The function to perform. If `test` represents a test suite,
+  ///     this function encapsulates running all the tests in that suite. If
+  ///     `test` represents a test function, this function is the body of that
+  ///     test function (including all cases if it is parameterized.)
+  ///   - test: The test under which `function` is being performed.
+  ///   - testCase: The test case, if any, under which `function` is being
+  ///     performed. When invoked on a suite, the value of this argument is
+  ///     `nil`.
+  ///
+  /// - Throws: Whatever is thrown by `function`, or an error preventing
+  ///   execution from running correctly. An error thrown from this method is
+  ///   recorded as an issue associated with `test`. If an error is thrown
+  ///   before `function` is called, the corresponding test will not run.
+  ///
+  /// When the testing library is preparing to run a test, it finds all traits
+  /// applied to that test (including those inherited from containing suites)
+  /// and asks each for its test executor (if any) by calling
+  /// ``Trait/executor(for:testCase:)-26qgm``. It then calls this method
+  /// on all non-`nil` instances, giving each an opportunity to perform
+  /// arbitrary work before or after invoking `function`.
+  ///
+  /// This method should either invoke `function` once before returning or throw
+  /// an error if it is unable to perform its custom logic successfully.
+  ///
+  /// This method is invoked once for the test its associated trait is applied
```

Comment:
I think this paragraph is now out of date? At a minimum, it should probably refer to the other function.